dual-stack, virtctl: expose ipv6 services

maiqueb opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
Adapt the virtctl API to accept the IPFamily of a the service to
be exposed; it accepts 'ipv4' and 'ipv6' as families.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add `ip-family` to the `virtctl expose` command. 
```

qinqon commented on pkg/virtctl/expose/expose.go:
```diff
@@ -69,6 +70,7 @@ virtualmachineinstance (vmi), virtualmachine (vm), virtualmachineinstancereplica
 	cmd.Flags().StringVar(&strTargetPort, "target-port", "", "Name or number for the port on the VM that the service should direct traffic to. Optional.")
 	cmd.Flags().StringVar(&strServiceType, "type", "ClusterIP", "Type for this service: ClusterIP, NodePort, or LoadBalancer.")
 	cmd.Flags().StringVar(&portName, "port-name", "", "Name of the port. Optional.")
+	cmd.Flags().StringVar(&strIPFamily, "ip-family", "ipv4", "IP family over which the service will be exposed.")
```
Add the valid values for example capitalized like IPv4 are not working.

maiqueb replied:
this is just the default; despite it being used also in the documentation, since we lower case it farther down the code, it won't matter much.

If you insist, I'll do it.

qinqon commented on pkg/virtctl/expose/expose.go:
```diff
@@ -125,6 +128,16 @@ func (o *Command) RunE(cmd *cobra.Command, args []string) error {
 
+	// convert from string to IPFamily type enum
+	switch strIPFamily {
+	case "ipv4":
```
what about using lowercase version of v1.IPv4Protocol ? and same for ipv6

maiqueb replied:
Not sure I understand you; I'l l just lowercase the variable, which seems fairer - the user's intentions if 'ipV4'  is type are clear.

qinqon replied:
I mean the following

```golang
switch strings.ToLower(v1.IPv4Protocol )
```

And yeah, it's easier to type it as lowercase.

qinqon reviewed (changes requested):
Looks good, small nit not args description.

maiqueb commented:
/cc @AlonaKaplan @phoracek

oshoval commented on tests/expose_test.go:
```diff
@@ -81,35 +97,104 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+			var serviceCleanupFunc func() error
+
+			BeforeEach(func() {
+				vmiExposeArgs = []string{
+					expose.COMMAND_EXPOSE,
+					"virtualmachineinstance", "--namespace", tcpVM.GetNamespace(), tcpVM.GetName(),
+					"--port", servicePort, "--name", serviceName,
+					"--target-port", strconv.Itoa(testPort),
+				}
```
can we put it under the `Context` of line 97 ?
since its the same always, 
or there was some reason (i think) to not create stuff directly under Context ?

oshoval replied:
I meant under the Context (which mean in the context) so it will still be on each context on its own.
but not crucial

phoracek replied:
I think this must be in BeforeEach since it must be called after the BeforeEach in outer context which sets the `tcpVM`

oshoval commented on tests/expose_test.go:
```diff
@@ -81,35 +97,104 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+			})
+
+			BeforeEach(func() {
+				serviceCleanupFunc = nil
```
please consider to merge the two nillers, into one, no real reason this time to split.

oshoval replied:
ye i know, but i dont see it a sufficient reason here, since they cant fail , but up to you

oshoval commented on tests/expose_test.go:
```diff
@@ -81,35 +97,104 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+			})
+
+			table.DescribeTable("[test_id:1531][label:masquerade_binding_connectivity]Should expose a Cluster IP service on a VMI and connect to it", func(ipFamily k8sv1.IPFamily) {
+				if ipFamily == k8sv1.IPv6Protocol {
```
maybe will be good to add test that also does it for ipv4 explicit, without considering it the default

oshoval replied:
i understand, but bugs can happen, lets see what other reviewers think ?

oshoval commented on tests/expose_test.go:
```diff
@@ -331,89 +615,166 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+			})
+
+			BeforeEach(func() {
+				Expect(createStoppedVM(virtClient, tests.NamespaceTestDefault)).To(Succeed(), "should have a stopped VM.")
```
"should create a stopped VM."
instead have i think

maiqueb replied:
Done.

oshoval commented on tests/expose_test.go:
```diff
@@ -331,89 +615,166 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+			AfterEach(func() {
+				if vm != nil {
+					By("Calling the stop command on the running VMI")
+					Expect(vm).NotTo(BeNil(), "should have a VM running so we can get rid of the VMI")
```
didnt we check it on line 691 ?
either leftovers or you want other check (or i am wrong)

oshoval commented on tests/expose_test.go:
```diff
@@ -425,8 +786,8 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
-					newUId := (vmiUIdAfterRestart != vmiUIdBeforeRestart)
-					return ((vmi.Status.Phase == v1.Running) && (newUId))
+					newUId := vmiUIdAfterRestart != vmiUIdBeforeRestart
+					return vmi.Status.Phase == v1.Running && newUId
```
i would merge the two lines

oshoval commented on tests/expose_test.go:
```diff
@@ -435,29 +796,48 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+			)
+
+			table.DescribeTable("[test_id:343][label:masquerade_binding_connectivity]Should Verify an exposed service of a VM is not functional after VM deletion.", func(ipFamily k8sv1.IPFamily) {
+				if ipFamily == k8sv1.IPv6Protocol {
```
lots of common code with line 744 ?
can we create some helper functions ?

oshoval replied:
yep, there is a block that appear around 3 times, just to find the biggest common snippet, add parameters if needed, and reduce code repetitions please

maiqueb replied:
Done.

maiqueb replied:
I take you mean 

```
				By("Calling the start command on the stopped VM")
				vmi := startVMIFromVMTemplate(vm.GetName(), vm.GetNamespace())
				Expect(vmi).NotTo(BeNil(), "should have been able to create a VMI from a VM")
				tests.GenerateHelloWorldServer(vmi, testPort, "tcp")
```

Those are 4 lines of code, out of which 1 is a `by` clause, and another an expectation. 

That leaves 2 lines of code, one that starts a VM, another that creates a server within it - with error handling in the middle. Encapsulating that in another function, imho, would be counter-productive since the function does not make much sense; what would you name it ? `startVMAndCreateAServerInIt` ? That 'and' in the middle is a clear indication it should be 2 functions, not one. 

Would you say the same if the code was

```golang
tests.GenerateHelloWorldServer(
    startVMIFromVMTemplate(vm.GetName(), vm.GetNamespace()),
    testPort,
    "tcp")
```
instead ?

oshoval replied:
> I take you mean
> 
> ```
> 				By("Calling the start command on the stopped VM")
> 				vmi := startVMIFromVMTemplate(vm.GetName(), vm.GetNamespace())
> 				Expect(vmi).NotTo(BeNil(), "should have been able to create a VMI from a VM")
> 				tests.GenerateHelloWorldServer(vmi, testPort, "tcp")
> ```
> 
> Those are 4 lines of code, out of which 1 is a `by` clause, and another an expectation.
> 
> That leaves 2 lines of code, one that starts a VM, another that creates a server within it - with error handling in the middle. Encapsulating that in another function, imho, would be counter-productive since the function does not make much sense; what would you name it ? `startVMAndCreateAServerInIt` ? That 'and' in the middle is a clear indication it should be 2 functions, not one.

`startVMWithServer`
but up to you if to do it or leave it as is.

> 
> Would you say the same if the code was
> 
> ```go
> tests.GenerateHelloWorldServer(
>     startVMIFromVMTemplate(vm.GetName(), vm.GetNamespace()),
>     testPort,
>     "tcp")
> ```
> 
> instead ?

nope, this one is self contained and reuse code.

maiqueb replied:
> `startVMWithServer`

Ah... Now it was not so hard ... 0_o 

I'll keep it as is since it allows me to do error handling in the test case body, where it belongs.

oshoval commented on tests/expose_test.go:
```diff
@@ -484,10 +864,57 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
-			})
+
+				// this way, the test does not attempt to stop the VM in the `AfterEach` section
+				vm = nil
```
maybe better to use defer, so even if there is a panic it would happen

oshoval replied:
right, so other question, having it in mind.
you create a new stopped vm for each test, right ?
so why to not stop / delete it on each end of a test ? 

generally speaking, you dont delete it , just stop it if all goes well, is it ok ?

maiqueb replied:
That means I need to create the VM once, and start / stop on the setup / teardown. I'd prefer to move as far away from `BeforeAll` as I can. 

Plus, creating the VM object is peanuts when compared to starting it.

oshoval replied:
if all goes well, who will stop the vm ?
each test would put nil and wont stop , nor delete it, or i miss something?

i find this reuse, even that it save time, little dangerous, and cause the code bit harder to maintain, we dont do it usually, this is my opinion at least.

maiqueb replied:
> if all goes well, who will stop the vm ?

If all goes well, the VM is deleted *within* the test. There's no VM to stop. 

> each test would put nil and wont stop , nor delete it, or i miss something?
> 
> i find this reuse, even that it save time, dangerous, and cause the code bit harder to maintain, we dont do it usually, this is my opinion at least.

> moreover, in case of a error in one of the test, you still create a new stopped vm, and there will be a leak of the previous vm, (it would be a stopped one, but still) right ?

Here you're right; error or not, the VM object is leaking. Gotta delete it.

maiqueb replied:
Done.

oshoval commented on tests/expose_test.go:
```diff
@@ -484,10 +864,57 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+		return "", err
+	}
+	for _, ipStr := range strings.Split(output, "\n") {
+		ip := strings.Split(ipStr, " ")[0]
```
aint space is the default delimiter ? 
unless you want to be explicit

oshoval replied:
i mixed with other languages where its overrides and have default parameters

oshoval reviewed (commented):
Nice

oshoval commented on tests/expose_test.go:
```diff
@@ -331,89 +616,175 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+
+			AfterEach(func() {
+				if vm != nil {
+					By("Deleting the VM ...")
```
so if we delete it , why we do need to stop it before that and waits its stopped?
lets just delete it ?

maiqueb replied:
Stopping the VM removes the VMI object.

Removing the VM removes the VM object.

Just deleting the VM object would leak the VMI object behind - leaving it in the failed state.

oshoval replied:
`oc delete vm ....`
would delete the vmi after a while isnt it ?
i do it all the time i think

maiqueb replied:
I think it eventually deletes it, but takes long to do so. 

Bottom line, we have to wait. Stopping the VM and waiting for the associated VMI to be removed is a 'graceful' shutdown, and as such, I prefer that approach to just deleting the VM.

oshoval replied:
do you want then to also wait after the VM delete so the vm would gone ?

maiqueb replied:
Might as well be thorough.

Done.

maiqueb commented on pkg/virtctl/expose/expose_test.go:
```diff
@@ -238,6 +238,27 @@ var _ = Describe("Expose", func() {
+			It("should succeed", func() {
+				err := tests.NewRepeatableVirtctlCommand(expose.COMMAND_EXPOSE, "vmi", vmName, "--name", "my-service",
+					"--port", "9999", "--target-port", "http", "--ip-family", "ipv4")
+				Expect(err()).To(BeNil())
```
Done.

oshoval commented on tests/expose_test.go:
```diff
@@ -484,10 +889,64 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
-			})
+
+				// this way, the test does not attempt to stop the VM in the `AfterEach` section
+				vm = nil
```
line 881
```
				By("Deleting the VM.")
				Expect(virtClient.VirtualMachine(vm.Namespace).Delete(vm.Name, &k8smetav1.DeleteOptions{})).To(Succeed())					Expect(virtClient.VirtualMachine(vm.Namespace).Delete(vm.Name, &k8smetav1.DeleteOptions{})).To(Succeed())
				tests.WaitForVirtualMachineToDisappearWithTimeout(vmi, 120)					tests.WaitForVirtualMachineToDisappearWithTimeout(vmi, 120)
```

1. u dont wait here for the vm to gone,
2. the vmi = nil should be put directly after that, else the assert below it would break,
the AfterEach will try to delete and assert again (but it will be silent and not that critical - just if you want to be more accurate)
3. you can use WaitForVirtualMachineToDisappearWithTimeout in your AfterEach where you "stop" maybe please

oshoval replied:
> > line 881
> > ```
> > 				By("Deleting the VM.")
> > 				Expect(virtClient.VirtualMachine(vm.Namespace).Delete(vm.Name, &k8smetav1.DeleteOptions{})).To(Succeed())					Expect(virtClient.VirtualMachine(vm.Namespace).Delete(vm.Name, &k8smetav1.DeleteOptions{})).To(Succeed())
> > 				tests.WaitForVirtualMachineToDisappearWithTimeout(vmi, 120)					tests.WaitForVirtualMachineToDisappearWithTimeout(vmi, 120)
> > ```
> > 
> > 
> > 
> > 1. u dont wait here for the vm to gone, (just vmi) - maybe will be good to remove this from the test itself and let the AfterEach do the job always ?
> 
> The goal of this test is to "Verify an exposed service of a VM is not functional after VM deletion". As such, the test must delete the VM.

It does delete the vmi, it doesnt wait for the vm to be gone as well, as you do in the AfterEach.

> 
> > 1. the vmi = nil should be put directly after that, else the assert below it would break,
> >    the AfterEach will try to delete and assert again (but it will be silent and not that critical - just if you want to be more accurate)
> 
> I think I understand what you mean; added that in a defer, so I always reset the pointer, no matter what happens in the test (only after the VM object is deleted).
> 
> > 1. you can use WaitForVirtualMachineToDisappearWithTimeout in your AfterEach where you "stop" maybe please
> 
> That function's input is a VMI, and we only have a VM. 'GET'ing the VMI from the VM in order to re-use that function (when I only want to cleanup) is going too far imo.

`vmi, err := virtClient.VirtualMachineInstance(vm.Namespace).Get(vm.Name, &k8smetav1.GetOptions{})`
as the test does, or whatever you want, just in order that the delete of the test would be equal to the delete of the AfterEach imo

maiqueb replied:
Added a when context that hopefully helps making this clearer; now there's a `When` clause for when just start / stop the VM, and another `When` clause when we just delete the VM .

Are you OK with this option ?

oshoval replied:
i dont understand why `When` or semantics is related here to be honestt,
i meant that there is asymmetric between the delete in the test versus the delete in the AfterEach,
the test doesnt wait for the VM itself to gone, where the AfterEach **did** i think ?

**did** - because where is it gone ? is it just me or its gone ? (where you delete the vmi in AfterEach in case of exception as it was before, (and i think you added also wait for vm to be gone).

moreoever i think that code reuse should be done as more as possible, but lets please first discuss where the AfterEach gone

maiqueb replied:
The delete on the test did not wait because it was the last test on a sequence of tests; once it finished, the next module would destroy the namespace.

The delete on the `AfterEach` needs to wait because there will be more tests executed afterwards (at least the delete test on ipv6 family), in which a new VM will have to be created.

There are now 2 when clauses:
  - 1 where the VM is created once; the VM is started in the tests - where the service is also exposed. The VM is then stopped in the test teardown. Since it is only stopped, there's no need to wait for the VM.
  - another where the VM is created in the test setup (since the test itself **deletes** the VM after it is gone.

Yes, I stopped waiting for the VM to be gone since it is either re-used, or we create a new one (this corner-case happens only on the last VM test, where it ends up deleting the VM).

Does it make sense now ?

oshoval replied:
imho i wouldnt do this split, it adds confusion, moreover

1. about `When("the VM object must be created")`
in case of expection, the vmi wont be deleted, since it wont reach the delete step, i prefer much the previous state that you did, where there is a guard using the AfterEach, even vmi isnt deleted here

2. `When("the VM object can be re-used")`
i am not sure we want to be depend on the next module (as you said in the first paragraph), and so on, unless its what there was before and considered non network test.
next module might be skipped and so on.

i personally find it confusing and not robust, i think each test should be self contained, at least to the extend it was before.
those assumptions make it harder to maintain

maiqueb replied:
OK, I see this has taken a turn for the worse. I'll get this back to it's initial state.

I'll change the test to wait everywhere since this seems to be upsetting you (thus aligning VM removal throughout the VM test).

This is the function I'm planning on using:

```golang
		cleanupVM := func(virtClient kubecli.KubevirtClient, namespace string, vmName string) {
			By("Deleting the VM ...")
			err := virtClient.VirtualMachine(vm.GetNamespace()).Delete(vm.GetName(), &k8smetav1.DeleteOptions{})
			Expect(err).NotTo(HaveOccurred(), "should have been able to delete the VM object")
			Eventually(func() bool {
				_, errVMI := virtClient.VirtualMachineInstance(vm.GetNamespace()).Get(vm.GetName(), &k8smetav1.GetOptions{})
				_, errVM := virtClient.VirtualMachine(vm.GetNamespace()).Get(vm.GetName(), &k8smetav1.GetOptions{})
				return errVMI != nil && errVM != nil && errors.IsNotFound(errVMI) && errors.IsNotFound(errVM)
			}, 120*time.Second, 1*time.Second).Should(BeTrue())
		}
```

oshoval replied:
Thanks, about the snippet, you can use `SatisfyAll(HaveOccurred(), WithTransform(errors.IsNotFound, BeTrue()))`
with two `Eventually` imo
and i guess that vm is always there when using it right ? (because of the first Expect)

oshoval replied:
Imo its more ginkgo oriented and modern than
`return errVMI != nil && errVM != nil && errors.IsNotFound(errVMI) && errors.IsNotFound(errVM)`

also its logical, you want the vmi to be gone, and then you want the vm to be gone,
no need to check again and again the vmi after its done, just because you wait vm to gone (i know it doesnt take a lot of time, still better logic and readability)

phoracek commented on tests/expose_test.go:
```diff
@@ -70,6 +74,18 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 	}
 
+	cleanupService := func(serviceName string, namespace string) func() error {
+		return func() error {
```
We need to block on the removal, otherwise we can fail to create another service with the same name while this one is being deleted. Although removal of a Service should be swift.

maiqueb replied:
Done.

phoracek commented on tests/expose_test.go:
```diff
@@ -70,6 +74,18 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+		}
+	}
+
+	cleanupJob := func(jobName string, namespace string) func() error {
```
ditto

maiqueb replied:
Done.

phoracek reviewed (changes requested):
Partial review before the weekend.

maiqueb commented on pkg/virtctl/expose/expose.go:
```diff
@@ -125,6 +128,17 @@ func (o *Command) RunE(cmd *cobra.Command, args []string) error {
+	case "ipv6":
+		ipFamily = v1.IPv6Protocol
+	default:
+		return fmt.Errorf("unknown IPFamily: %s", strIPFamily)
```
None of the other options do that, which would make this behavior inconsistent.

phoracek replied:
It would make the behavior superior. Us look smarter and the user happier.

(Can't believe I'm saying this) I don't see how consistency here matters that much. All errors are including a string, this will be slightly more verbose. But maybe there are issues - the list getting out of sync?

maiqueb replied:
I got no issue against happier users; I got an issue with trying to piggy back that in this PR.

It will be done as a follow-up pr. OK with you ?

phoracek replied:
Totally, thanks!

alonSadan commented on tests/expose_test.go:
```diff
@@ -81,34 +94,96 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 				By("Getting back the cluster IP given for the service")
 				svc, err := virtClient.CoreV1().Services(tcpVM.Namespace).Get(serviceName, k8smetav1.GetOptions{})
 				Expect(err).ToNot(HaveOccurred())
+				serviceCleanupFunc = cleanupService(serviceName, tcpVM.GetNamespace())
```
Is it possiable to move this to the beforeEach() (or maybe the afterEach()?) ?  I think it is more suitable there cause it is not so related to the test itself, but more of an infra thing. maybe instead of :
```
BeforeEach(func() {
    serviceCleanupFunc = nil
})
```
you can: 
```
BeforeEach(func() {
    serviceCleanupFunc = cleanupService(serviceName, tcpVM.GetNamespace())
})

alonSadan replied:
Than maybe in the afterEach()?
why it's important to keep it in the test body?

maiqueb replied:
> Right. Then maybe in the afterEach()?

That would be possible, but I don't see what's the point. I'm creating the service in the test, and storing the pointer to it also on the test. Then,  the test tear-down just calls the cleanup function. 

What you say would require me to store a pointer to the service, and then I'd have to use this to delete it. What are the advantages ? What are you trying to prevent on the current approach ? 

> why it's important to keep it in the test body?

The order of the tests (as described in our functional test plan) is:
  - create VM
  - expose service from stopped VM
  - start VM
  - check connectivity

We are creating the tests on a table, having the IP family as the parameter. Accessing the IP family is only possible **in** the tests, not in the setup, nor tear-down of the test.

This limits us to expose the service *within* the tests. If we want to preserve what the test does (I think we should) we can create the VM on the test setup, but are forced to expose the service & start the VM in the test code.

maiqueb replied:
> > > Right. Then maybe in the afterEach()?
> > 
> > 
> > That would be possible, but I don't see what's the point. I'm creating the service in the test, and storing the pointer to the callback that deletes it also on the test. Then, the test tear-down just calls the cleanup function.
> > What you say would require me to store a pointer to the service, and then I'd have to use this to delete it. What are the advantages ? What are you trying to prevent on the current approach ?
> > > why it's important to keep it in the test body?
> > 
> > 
> > The order of the tests (as described in our functional test plan) is:
> > 
> > * create VM
> > * expose service from stopped VM
> > * start VM
> > * check connectivity
> > 
> > We are creating the tests on a table, having the IP family as the parameter. Accessing the IP family is only possible **in** the tests, not in the setup, nor tear-down of the test.
> > This limits us to expose the service _within_ the tests. If we want to preserve what the test does (I think we should) we can create the VM on the test setup, but are forced to expose the service & start the VM in the test code.
> 
> I think I get what I missed here- that it is a table and the afterEach won't run after each entry, so you had to use the callback. am I right?

What ? No, the table entries generate 2 tests, which have setup / tear-down.

> My point was to keep things simpler, avoid using callbacks, and keep the tests without teardown-related processes, but it is not that crucial I think.
> Thanks for explaining in such details.

Cleaning stuff up in the tear-down is the proper way to handle this.

From where I stand, it seems you don't like the callback; I'm OK with that, if you insist, I can change it to be a regular function call .

> 
> By the way, in order to run beforeEach() and afterEach() after each table entry it is possible to do something like suggested here:
> [onsi/ginkgo#378 (comment)](https://github.com/onsi/ginkgo/issues/378#issuecomment-501538323)
> but again, not crucial, and thanks for explaining.

alonSadan commented on tests/expose_test.go:
```diff
@@ -81,34 +94,96 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 
 				By("Starting a job which tries to reach the VMI via ClusterIP")
 				job := runHelloWorldJob(serviceIP, servicePort, tcpVM.Namespace)
+				jobCleanupFunc = cleanupJob(job.GetName(), job.GetNamespace())
```
I would say that as for `serviceCleanupFunc` the same applies for `jobCleanupFunc`, but here the names are not consts or vars so it will be harder. consider moving this to beforeEach() as well, but not a must.

alonSadan commented on tests/expose_test.go:
```diff
@@ -81,34 +94,96 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+				By("Getting back the service so we can clean it up later on")
+				service, err := virtClient.CoreV1().Services(tcpVM.Namespace).Get(serviceName, k8smetav1.GetOptions{})
 				Expect(err).ToNot(HaveOccurred())
+				serviceCleanupFunc = cleanupService(service.GetName(), service.GetNamespace())
```
ditto

alonSadan commented on tests/expose_test.go:
```diff
@@ -152,24 +256,57 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 				By("Getting back the service")
 				svc, err := virtClient.CoreV1().Services(tcpVM.Namespace).Get(serviceName, k8smetav1.GetOptions{})
 				Expect(err).ToNot(HaveOccurred())
+				serviceCleanupFunc = cleanupService(svc.GetName(), svc.GetNamespace())
```
ditto

alonSadan commented on tests/expose_test.go:
```diff
@@ -203,42 +384,114 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 				By("Getting back the cluster IP given for the service")
 				svc, err := virtClient.CoreV1().Services(udpVM.Namespace).Get(serviceName, k8smetav1.GetOptions{})
 				Expect(err).ToNot(HaveOccurred())
+				serviceCleanupFunc = cleanupService(svc.GetName(), svc.GetNamespace())
```
ditto

alonSadan commented on tests/expose_test.go:
```diff
@@ -203,42 +384,114 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 				By("Getting back the cluster IP given for the service")
 				svc, err := virtClient.CoreV1().Services(udpVM.Namespace).Get(serviceName, k8smetav1.GetOptions{})
 				Expect(err).ToNot(HaveOccurred())
+				serviceCleanupFunc = cleanupService(svc.GetName(), svc.GetNamespace())
```
ditto

alonSadan commented on tests/expose_test.go:
```diff
@@ -331,89 +637,185 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 				By("Getting back the cluster IP given for the service")
 				svc, err := virtClient.CoreV1().Services(vm.Namespace).Get(serviceName, k8smetav1.GetOptions{})
 				Expect(err).ToNot(HaveOccurred())
+				serviceCleanupFunc = cleanupService(svc.GetName(), svc.GetNamespace())
```
ditto

alonSadan commented on tests/expose_test.go:
```diff
@@ -331,89 +637,185 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 				By("Getting back the service's allocated cluster IP.")
 				svc, err := virtClient.CoreV1().Services(vmObj.Namespace).Get(serviceName, k8smetav1.GetOptions{})
 				Expect(err).ToNot(HaveOccurred())
+				serviceCleanupFunc = cleanupService(svc.GetName(), svc.GetNamespace())
```
ditto

alonSadan commented on tests/expose_test.go:
```diff
@@ -305,24 +573,62 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 				By("Getting back the cluster IP given for the service")
 				svc, err := virtClient.CoreV1().Services(vmrs.Namespace).Get(serviceName, k8smetav1.GetOptions{})
 				Expect(err).ToNot(HaveOccurred())
+				serviceCleanupFunc = cleanupService(svc.GetName(), svc.GetNamespace())
```
ditto

alonSadan reviewed (commented):
Thanks Miguel. I added one comment that repeated so I dittoed it, sorry for the noise. 
I got a bit confused by the lambdas, and I think that if you accept that `serviceCleanupFunc`
doesn't have to be in the tests it self, then maybe it is also possiable to let go of the lambda and do : 
```
serviceCleanupFunc := func(serviceName string, namespace string) error { 
return virtClient.CoreV1().Services(namespace).Delete(serviceName, &k8smetav1.DeleteOptions{})
			Eventually(func() error {
				_, err := virtClient.CoreV1().Services(namespace).Get(serviceName, k8smetav1.GetOptions{})
				return err
			}, 2*time.Minute, time.Second).Should(SatisfyAll(HaveOccurred(), WithTransform(errors.IsNotFound, BeTrue())), "The `Corev1.Service` entity should be gone within the given timeout")
			return err
		}
	}
```

alonSadan reviewed (commented):
Reviewed again. Added one question. Other than that looks very good:)
I think I'll have to make another round just to be sure before lgtm, cause it is a long PR

maiqueb commented on tests/expose_test.go:
```diff
@@ -81,34 +107,104 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+				Expect(virtctl()).To(Succeed(), "should expose a service via `virtctl expose ...`")
+
+				By("Getting back the service so we can clean it up later on")
+				service, err := virtClient.CoreV1().Services(tcpVM.Namespace).Get(serviceName, k8smetav1.GetOptions{})
```
I think I'm now doing it.

maiqueb commented:
> Reviewed again. Added one question. Other than that looks very good:)
> I think I'll have to make another round just to be sure before lgtm, cause it is a long PR.
> 
> about the callbacks, I won't insist, whatever you prefer.

I've pushed a version where I removed the callbacks, let's see if this is clearer.

Also moved to the previous solution without the `When` clauses.

alonSadan commented on tests/expose_test.go:
```diff
@@ -70,6 +76,35 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 
+	cleanupService := func(serviceName string, namespace string) error {
+		err := virtClient.CoreV1().Services(namespace).Delete(serviceName, &k8smetav1.DeleteOptions{})
+		err = wait.PollImmediate(2*time.Minute, time.Second, func() (done bool, err error) {
```
nice!

alonSadan commented:
> > Reviewed again. Added one question. Other than that looks very good:)
> > I think I'll have to make another round just to be sure before lgtm, cause it is a long PR.
> > about the callbacks, I won't insist, whatever you prefer.
> 
> I've pushed a version where I removed the callbacks, let's see if this is clearer.
> 
> Also moved to the previous solution without the `When` clauses.

I really liked the new version:) I think it is more understandable. Thank you for this PR.
/lgtm

qinqon commented on tests/expose_test.go:
```diff
@@ -474,8 +797,9 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 				By("Deleting the VM.")
-				Expect(virtClient.VirtualMachine(vm.Namespace).Delete(vm.Name, &k8smetav1.DeleteOptions{})).To(Succeed())
-				tests.WaitForVirtualMachineToDisappearWithTimeout(vmi, 120)
+				Expect(cleanupVM(virtClient, vm.GetNamespace(), vm.GetName())).To(Succeed(), "should have been able to delete the VM object")
+				// this way, the test does not attempt to stop the VM in the `AfterEach` section
+				defer func() { vm = nil }()
```
better if we just make the code at `AfterEach` to discover if it has to stop it or not.

maiqueb replied:
what's the advantage ?

maiqueb replied:
> what's the advantage ? It costs an extra read just to know if it must get rid of it.

qinqon replied:
This is implementing cleanup code at an It

oshoval replied:
> This is implementing cleanup code at an It

This is also what i suggested, in one of the threads, combine to removals, put them just on the AfterEach.
We can indeed try to delete the vm, if the delete has an error which isnt isnot found - assert, 
if its is not found - all good, if its found - wait for deletion of vmi, vm as you do now.

oshoval replied:
> This is implementing cleanup code at an It

In case it doesnt hurt the test logic, 
this is also what i voted in one of the threads, combine the removals.
You can do something like
if vm != nil (you must, else you might deref nil),
delete - if error == isnotfound - all good
if error != nil , != isnotfound - assert
else wait for all gone

maiqueb replied:
> This is implementing cleanup code at an It

If this is cleanup code, what about removing the VM ? That's also cleanup code. But it's part of the test. 

My point is that those 2 things (deleting object / re-setting pointer) belong together.

qinqon replied:
ok, fair enough

qinqon reviewed (changes requested):
LGTM but let's see if we can remove that ugly defer.

oshoval commented on tests/expose_test.go:
```diff
@@ -474,8 +797,9 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
-				tests.WaitForVirtualMachineToDisappearWithTimeout(vmi, 120)
+				Expect(cleanupVM(virtClient, vm.GetNamespace(), vm.GetName())).To(Succeed(), "should have been able to delete the VM object")
+				// this way, the test does not attempt to stop the VM in the `AfterEach` section
+				defer func() { vm = nil }()
```
imo its better to write 
```
defer func() {
 ...
}()
```

maiqueb replied:
In mine it isn't, and the compiler / linter / syntax checker ate it, so it's legit.

oshoval commented:
Thanks
/lgtm

oshoval commented on tests/expose_test.go:
```diff
@@ -331,89 +557,172 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+			})
+
+			BeforeEach(func() {
+				Expect(createStoppedVM(virtClient, tests.NamespaceTestDefault)).To(Succeed(), "should create a stopped VM.")
```
i would suggest to add vm = nil as the first line
so it wont remember the vm which isnt nil from the previous cycle, in case the only line there fails.

maiqueb replied:
Why should I care about the pointer in the before each ?

oshoval replied:
because you depend on it in the AfterEach and if the BeforeEach asserted, you would still have the previous value, and fail again on the assert of Delete in the AfterEach, since the ptr isnt nil but the vm is already deleted

oshoval commented on tests/expose_test.go:
```diff
@@ -331,89 +557,172 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+
+			err = wait.PollImmediate(time.Second, 2*time.Minute, func() (done bool, err error) {
+				_, err = virtClient.VirtualMachineInstance(vm.GetNamespace()).Get(vm.GetName(), &k8smetav1.GetOptions{})
+				return err != nil && errors.IsNotFound(err), nil
```
dont u need to return err as the 2nd arg here ?

maiqueb replied:
no, not at all; see https://github.com/kubernetes/apimachinery/blob/master/pkg/util/wait/wait.go#L430

It returns when it finds an error.

oshoval replied:
ok so it would swallow the real error, and we would have just a custom error of PollImmediate right?

maiqueb replied:
This is a test helper, and as such, I'd rather not rely on Eventually, or for that matter, any other type of ginkgo stuff if I have a sound alternative. In this case, I do.

Furthermore, why do you want the error ? What would you use it for ? I just want to check that I managed to get rid of the k8s entity. Let's please keep on track on what this function is supposed to do here please.

oshoval replied:
since if it fails, it adds info,
i less agree with the both sections you said above, but up to you

oshoval commented on tests/expose_test.go:
```diff
@@ -474,8 +797,9 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
-				tests.WaitForVirtualMachineToDisappearWithTimeout(vmi, 120)
+				Expect(cleanupVM(virtClient, vm.GetNamespace(), vm.GetName())).To(Succeed(), "should have been able to delete the VM object")
+				// this way, the test does not attempt to stop the VM in the `AfterEach` section
+				defer func() { vm = nil }()
```
you can just `vm = nil` , without defer here, no need to add boilerplate since after this line it got to go directly

maiqueb commented:
/retest

Read the VM to know if it must be removed in the test tear-down.

qinqon commented:
/lgtm

maiqueb commented:
/retest

oshoval commented on tests/expose_test.go:
```diff
@@ -331,89 +557,177 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+			})
+
+			AfterEach(func() {
+				if doesVMExist(vm.GetName(), vm.GetNamespace()) {
```
this vm.Name deref is buggy,
if the vm doesnt exist its a nil ptr deref
for example if 
https://github.com/kubevirt/kubevirt/pull/4272/files#diff-bf4f2c831a0b99974eedbd27880b2ebc7d3e68e63a5a06613be323930225d1d6R642 is failed

maiqueb replied:
If that fails, we don't have a VM, and the test is already failed.

Meaning there's nothing to cleanup, and nothing to test. From the perspective of the test framework, there's nothing to do.

oshoval replied:
but there is a panic, silent one but it exists, agree?
i am not sure we are fine with silent panics even that they are silent

maiqueb replied:
I agree that in the case you present it will error - but we will **never** see the error. The test will already be failed, and there is nothing to cleanup.

I really don't understand why you want to add more code to control a scenario that you will never face (the error is eaten away, and the real error presented - i.e. why did the VM fail to be created in the first place).

oshoval replied:
just because we might want the code to not swallow errors as this one, especially when the solution is easy imo, 
setting the vm to nil and checking if its nil before deref, this is the right thing to do imo

maiqueb replied:
It is not us that are swallowing the error, it is the test framework - which is specified to work like that. 

You keep pointing out that "the solution is easy". And I keep pointing out that there is no problem to fix. Unless you come up with a real reason why there is a problem, the fact that the solution is easy is irrelevant (since there is no problem).

oshoval replied:
masking the error is a problem imo, one might consider it minor, one can consider it not problem at all,
i do consider it a minor problem, and when i see there is an easy and right solution, cleaning a variable that might be used wrongly, the problem gets little bigger in my perspective

up to you what to do, this is my opinion

maiqueb replied:
> masking the error is a problem imo, one might consider it minor, one can consider it not problem at all,

Once again: I am not masking any error. The test framework is. You're free to open a bug to ginkgo explaining you think they're masking errors, and asking their behavior to change. 

> i do consider it a minor problem, and when i see there is an easy and right solution, cleaning a variable that might be used wrongly, the problem gets little bigger in my perspective
> 
> up to you what to do, this is my opinion

You're entitled to it.

oshoval replied:
> > masking the error is a problem imo, one might consider it minor, one can consider it not problem at all,
> 
> Once again: I am not masking any error. The test framework is. You're free to open a bug to ginkgo explaining you think they're masking errors, and asking their behavior to change.

We still can / might want to code such it wont happen, generally speaking.

> 
> > i do consider it a minor problem, and when i see there is an easy and right solution, cleaning a variable that might be used wrongly, the problem gets little bigger in my perspective
> > up to you what to do, this is my opinion
> 
> You're entitled to it.

maiqueb commented:
/cc @AlonaKaplan 

I think this should be reviewed as is, since the tests currently create the VMI once.

Destroying the VMIs and re-creating them makes the tests take > 10 additional minutes (just from +- 17 minutes to +- 28 minutes) .

AlonaKaplan commented on pkg/virtctl/expose/expose.go:
```diff
@@ -69,6 +70,7 @@ virtualmachineinstance (vmi), virtualmachine (vm), virtualmachineinstancereplica
 	cmd.Flags().StringVar(&strTargetPort, "target-port", "", "Name or number for the port on the VM that the service should direct traffic to. Optional.")
 	cmd.Flags().StringVar(&strServiceType, "type", "ClusterIP", "Type for this service: ClusterIP, NodePort, or LoadBalancer.")
 	cmd.Flags().StringVar(&portName, "port-name", "", "Name of the port. Optional.")
+	cmd.Flags().StringVar(&strIPFamily, "ip-family", "IPv4", "IP family over which the service will be exposed. Valid values are 'IPv4' or 'IPv6'.")
```
I would remove the "IPv" prefix, I think `--ip-family=4` `--ip-family=6` is enough.

maiqueb replied:
Since the attribute is configured w/ `--ip-family` I'd prefer to stick w/ `IPv4` /  `IPv6`, since it matches the possible values of the k8s `IPFamily` attribute.

https://github.com/kubevirt/kubevirt/blob/d072d7ffcbba8d862e79db7749fd0effe2ca2f4e/vendor/k8s.io/api/core/v1/types.go#L3824

AlonaKaplan replied:
I have to admit Miguel convinced me. We have to be aligned with k8s services.

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -491,3 +494,40 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+	}
+	for _, ipStr := range strings.Split(output, "\n") {
+		ip := strings.Split(ipStr, " ")[0]
+		if libnet.GetFamily(ip) == ipFamily {
```
ahostsv4/ahostsv6 can be used instead.

maiqueb replied:
Alright ! Good catch, my ignorant past self was unaware that existed :) 

Thanks.

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -84,34 +116,72 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+			})
+
+			AfterEach(func() {
+				Expect(job).NotTo(BeNil(), "a successful test must have stored a way to delete the batchv1.Job entity")
```
The job has [TTLSecondsAfterFinished](https://github.com/kubevirt/kubevirt/blob/d380576d2937b86fcd2cff8a73976fd32186a02e/tests/job.go#L99), afaiu, it means it will be automatically cleaned up.

Do you think this mechanism is not working and we have a leak?

maiqueb replied:
It does have, but is defined to 60 seconds. As a result, some of the faster tests will result in a pile of `Job`s.

I'll confirm this TTL mechanism works, and add a method that allows us to specify the TTL, setting it to 0. 

Do you happen to know a reason why we should use a TTL != 0 ? (delete instantly after it fails / completes)

AlonaKaplan replied:
I guess to collect logs.

maiqueb replied:
We do not collect the `Job`s logs: https://github.com/kubevirt/kubevirt/blob/4a7174ef8907f6eff8e3362bde4fc67030605b38/tests/reporter/kubernetes.go#L103

.... but that doesn't mean we never will. 

Should I just use this version w/ a fixed 60 second TTL ?

AlonaKaplan replied:
Since that's how it worked till now, and I'm not aware of any issues with it. I would keep it with the 60 seconds ttl.

maiqueb replied:
I agree with this, and right now, we do not even store the logs of jobs.

I personally don't think that there's much chance of the job log actually having anything meaningful / useful for debug - at least the ones we have right now (given their implementation) .

I would personally just delete them, without any regard for their logs.

I can also just follow the current status quo, and do nothing with them, and they will go away on their own, since each alternative has its advantages (doing nothing vs being explicit + making sure there are no leaks) . 

Which of the 2 should I implement ? @EdDev / @AlonaKaplan

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -73,6 +76,35 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+		err := virtClient.CoreV1().Services(namespace).Delete(serviceName, &k8smetav1.DeleteOptions{})
+		err = wait.PollImmediate(2*time.Minute, time.Second, func() (done bool, err error) {
+			_, err = virtClient.CoreV1().Services(namespace).Get(serviceName, k8smetav1.GetOptions{})
+			return err != nil && errors.IsNotFound(err), nil
```
Why do we need to wait for the service to be removed? Isn't it possible to add a random string to the service name?

maiqueb replied:
We can.

Waiting for it to be gone is easy, and takes a very short amount of time.

I find it simpler, and easier to know before-hand what is the name of the service when debugging; it saves me the operation of going to check the real service name.

If you insist I can do it like you suggest.

AlonaKaplan replied:
The prefix of the service name will still be the same, so for the purpose of debugging I don't see a big difference. Maybe it will be even easier to read the logs since different test will have different suffix for the service name.

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -84,34 +116,72 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
-					tcpVM.Namespace, tcpVM.Name, "--port", servicePort, "--name", serviceName, "--target-port", strconv.Itoa(testPort))
-				err := virtctl()
-				Expect(err).ToNot(HaveOccurred())
+				virtctl := tests.NewRepeatableVirtctlCommand(getVirtctlExposeCommand(vmiExposeArgs, ipFamily)...)
```
Missing
```
if family == k8sv1.IPv6Protocol {
				libnet.SkipWhenNotDualStackCluster(virtClient)
			}
```

maiqueb replied:
Sorry for misleading you.

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -84,34 +116,72 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+				Expect(cleanupService(serviceName, tcpVM.GetNamespace())).To(Succeed(), "should be able to delete the k8sv1.Service entity")
+			})
+
+			table.DescribeTable("[test_id:1532]Should expose a ClusterIP service and connect to the vm on port 80", func(ipFamily k8sv1.IPFamily) {
```
I would move the "[test_id:1532]" to the ipv4 entry only. We don't know if the corresponding test covers ipv6 as well. The qe guys should go over it and update the test-ids if it is.

AlonaKaplan replied:
same regarding the missing skip.

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -127,16 +197,33 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+				Expect(cleanupService(serviceName, tcpVM.GetNamespace())).To(Succeed(), "should be able to delete the k8sv1.Service entity")
+			})
+
+			table.DescribeTable("[test_id:1533]Should expose a ClusterIP service and connect to all ports defined on the vmi", func(ipFamily k8sv1.IPFamily) {
```
same regarding the "test-id"

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -155,20 +242,42 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+				Expect(cleanupService(serviceName, tcpVM.GetNamespace())).To(Succeed(), "should be able to delete the k8sv1.Service entity")
+			})
+
+			table.DescribeTable("[test_id:1534[label:masquerade_binding_connectivity]Should expose a NodePort service on a VMI and connect to it", func(ipFamily k8sv1.IPFamily) {
```
same regarding the "test-id" and the skip.

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -184,13 +293,28 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 					nodeIP := node.Status.Addresses[0].Address
 
+					if ipFamily == k8sv1.IPv6Protocol {
+						ipv6NodeIP, err := resolveNodeIPAddrByFamily(
```
According to the [documentation](https://kubernetes.io/docs/tasks/network/validate-dual-stack/#validate-node-addressing) the `node.Status.Addresses` should contain both `ipv4` and `ipv6` addresses. 

IIUC, [this](https://github.com/kubernetes/kubernetes/pull/95239/files#diff-15a7ea42c856468667780447090b682f36ea0385dcdbe9105ff49900564a190bR101) merged PR is solving the issue.

@oshoval can we please update the cluster?

EdDev replied:
@AlonaKaplan , I do not know if that PR is relevant to this issue, they did spoke about pod HostIP while the function handled the node.
On k8s, there was a long standing test that covered this, see [here](https://github.com/kubernetes/kubernetes/commit/789920955a9cda382d1d4d5557469c881d8c93f7#diff-20f023ec27befdd70ec0f9283ebabd364679d9ab8b7f95ae87449159d58396a5R59).
So I am not clear why the two addresses are not seen today or how that test is passing on k8s.

Perhaps we are not setting or defining the nodes when registering them to the cluster in a way that the dual-stack is detected right?

EdDev replied:
I think that if the [cloud-controller-manager](https://kubernetes.io/docs/concepts/architecture/cloud-controller/) is in place, the two addresses are seen. For bare metal, the ref PR may be a solution, I do not really understand (as it mentions passing the IP addresses manually in the config, which is weird) if this is what is really missing. Seems to me that calico also needs to be in the picture here.

maiqueb replied:
There it is: https://github.com/kubevirt/kubevirtci/issues/462

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -84,34 +116,72 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+				Expect(cleanupService(serviceName, tcpVM.GetNamespace())).To(Succeed(), "should be able to delete the k8sv1.Service entity")
+			})
+
+			table.DescribeTable("[test_id:1531][label:masquerade_binding_connectivity]Should expose a Cluster IP service on a VMI and connect to it", func(ipFamily k8sv1.IPFamily) {
```
I would move the test id to the ipv4 entry.

maiqueb replied:
Done.

AlonaKaplan commented on tests/job.go:
```diff
@@ -140,7 +140,7 @@ func NewHelloWorldJobUDP(host string, port string) *batchv1.Job {
 // On success - it expects to receive "Hello World!".
 func NewHelloWorldJobHTTP(host string, port string) *batchv1.Job {
-	check := []string{fmt.Sprintf(`set -x; x="$(head -n 1 < <(curl %s:%s))"; echo "$x" ; if [ "$x" = "Hello World!" ]; then echo "succeeded"; exit 0; else echo "failed"; exit 1; fi`, FormatIPForURL(host), port)}
+	check := []string{fmt.Sprintf(`set -x; ping -c 1 %s; x="$(head -n 1 < <(curl %s:%s))"; echo "$x" ; if [ "$x" = "Hello World!" ]; then echo "succeeded"; exit 0; else echo "failed"; exit 1; fi`, host, FormatIPForURL(host), port)}
```
You wrote in the commit message the problem is just for `nc`? Do we have the same issue with curl?

maiqueb replied:
The problem does not reproduce if you open a shell in the job and nc / curl manually. That was what lead me into using this silly work-around.

AlonaKaplan commented on tests/job.go:
```diff
@@ -129,8 +129,8 @@ func NewHelloWorldJobUDP(host string, port string) *batchv1.Job {
 	localPort--
-	check := []string{fmt.Sprintf(`set -x; trap "kill 0" EXIT; x="$(head -n 1 < <(echo | nc -up %d %s %s -i 3 -w 3 & nc -ul %d))"; echo "$x" ; if [ "$x" = "Hello UDP World!" ]; then echo "succeeded"; exit 0; else echo "failed"; exit 1; fi`,
-		localPort, host, port, localPort)}
+	check := []string{fmt.Sprintf(`set -x; ping -c 1 %s; trap "kill 0" EXIT; x="$(head -n 1 < <(echo | nc -up %d %s %s -i 3 -w 3 & nc -ul %d))"; echo "$x" ; if [ "$x" = "Hello UDP World!" ]; then echo "succeeded"; exit 0; else echo "failed"; exit 1; fi`,
```
General comment - 
1. If the issue is ipv6 one, can we add the ping only to the ipv6 cases?
2. Is there an open issue about the bug? 
IIUC, before using an ipv6 service the users will have to send a request with no port to the service before being able to send a request with a port.
Sounds like a bug that we should document. Does regular pods + k8s services have the same issue as well?

maiqueb replied:
> General comment -
> 
> 1. If the issue is ipv6 one, can we add the ping only to the ipv6 cases?

Yes, but that would require me to update these functions with the ip family, as they currently do not care at all about it.

I would rather ping always (simpler code, less changes, it's what we already do for `NewHelloWorldJob`).

> 2. Is there an open issue about the bug?

No. I'll create one.

>    IIUC, before using an ipv6 service the users will have to send a request with no port to the service before being able to send a request with a port.

Could be, I did not search deep enough.

>    Sounds like a bug that we should document. Does regular pods + k8s services have the same issue as well?

What do you mean by regular ? IPv4 ? If so, no. They do not require it.

AlonaKaplan replied:
> > General comment -
> > 
> > 1. If the issue is ipv6 one, can we add the ping only to the ipv6 cases?
> 
> Yes, but that would require me to update these functions with the ip family, as they currently do not care at all about it.
> 
> I would rather ping always (simpler code, less changes, it's what we already do for `NewHelloWorldJob`).

I think we should fix `NewHelloWorldJob` as well. It is a bug and it should be clear it is a workaround for ipv6 only.

> 
> > 1. Is there an open issue about the bug?
> 
> No. I'll create one.
> 
> > IIUC, before using an ipv6 service the users will have to send a request with no port to the service before being able to send a request with a port.
> 
> Could be, I did not search deep enough.
> 
> > Sounds like a bug that we should document. Does regular pods + k8s services have the same issue as well?
> 
> What do you mean by regular ? IPv4 ? If so, no. They do not require it.

No, I don't mean ipv4. I mean running the tcp/udp/http server on a pod, not a vm. And creating the NodePort service using the k8s api and not `virtctl expose`.

AlonaKaplan replied:
@maiqueb please answer the [comment](https://github.com/kubevirt/kubevirt/pull/4272/commits/370860998bf1853d80fa151b2735b44ad354a38f#r511774774).

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -330,38 +330,79 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+				Expect(cleanupService(serviceName, udpVM.GetNamespace())).To(Succeed(), "should be able to delete the k8sv1.Service entity")
+			})
+
+			table.DescribeTable("[test_id:1535][label:masquerade_binding_connectivity]Should expose a ClusterIP service on a VMI and connect to it", func(ipFamily k8sv1.IPFamily) {
```
Please move the test id to the ipv4 entry and skip the ipv6 test in case the cluster is not dual stack. Relevant to all the tests on the commit.

AlonaKaplan replied:
same as https://github.com/kubevirt/kubevirt/pull/4272/commits/b3f10f3697674f74bd8763385ef93ec16d4027cf#r511586890

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -860,3 +844,19 @@ func getVirtctlExposeCommand(currentArgs []string, ipFamily k8sv1.IPFamily) []st
+	if err != nil {
+		return "", err
+	}
+	return svc.Spec.ClusterIP, err
```
s/err/nil

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -860,3 +844,19 @@ func getVirtctlExposeCommand(currentArgs []string, ipFamily k8sv1.IPFamily) []st
+	if err != nil {
+		return 0, err
+	}
+	return int(svc.Spec.Ports[0].NodePort), err
```
s/err/nil

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -428,11 +424,10 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+				Expect(err).ToNot(HaveOccurred(), "should have been able to retrieve the service IP")
+
+				nodePort, err := getServiceNodePort(virtClient, udpVM.GetNamespace(), serviceName)
 				Expect(nodePort).To(BeNumerically(">", 0))
```
Expect(err).ToNot(HaveOccurred(), "should have been able to retrieve the node port")

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -603,6 +603,36 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 			return vmi
 		}
 
+		cleanupVM := func(virtClient kubecli.KubevirtClient, namespace string, vmName string) error {
```
Besides tests that actually test deleting the vm. I believe the vm cleanup is not needed and we can use `BeforeTestCleanup`.

EdDev replied:
Just make sure we do delete there everything (jobs, services, etc...)

EdDev replied:
Passing in the vm object itself makes more sense to me (we are deleting a known object).
But the real problem is that you pass a `namespace` and `vmName` but never use them.

Please do not use anonymous functions if it is not a must, everytime I see them I am looking for the reason they exist, i.e. the variable that is shared with the outside).

maiqueb commented on tests/expose_test.go:
```diff
@@ -633,19 +626,19 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 
 			AfterEach(func() {
-				Expect(cleanupService(serviceName, vm.GetNamespace())).To(Succeed(), "should be able to delete the k8sv1.Service entity")
+				Expect(cleanupService(serviceName, tests.NamespaceTestDefault)).To(Succeed(), "should be able to delete the k8sv1.Service entity")
```
Because one of the VM tests (the last one) deletes the vm. 

As such, I don't have a VM to get the name from.

I can add a variable to store the name instead.

AlonaKaplan replied:
I think this `AfterEach` is redundant anyway since as I mentioned in other places, I would move `tests.BeforeTestCleanup()` to `beforeEach` and remove the `AfterEach`es.

AlonaKaplan replied:
Yes, the comment was about the vm related tests only.

maiqueb replied:
Removed all the `AfterEach` clauses on the VM related tests.

"Resolving" this.

maiqueb commented on tests/expose_test.go:
```diff
@@ -536,47 +536,99 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+				virtctl := tests.NewRepeatableVirtctlCommand(getVirtctlExposeCommand(vmExposeArgs, ipFamily)...)
+				Expect(virtctl()).ToNot(HaveOccurred(), "should succeed exposing a service via `virtctl expose ...`")
+
+				vmi := startVMWithServer("tcp", testPort)
```
Because in the VM tests, in one of the tests I have to create a new VM (since the last test effectively deletes the VM object).

Since I have to re-create it once, might as well do it more times (I've tried to play around with the contexts - but I didn't like how it read in the end).

maiqueb commented on tests/expose_test.go:
```diff
@@ -536,47 +536,99 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+			})
+
+			AfterEach(func() {
+				if vm != nil {
```
I've tried an approach relying on calling `BeforeTestCleanup` and it bloats the test execution time by > 10 minutes.

I would avoid it for that reason.

AlonaKaplan replied:
Why? You don't have any shared resources between the tests.

maiqueb replied:
I'm guessing creating storage for the VMI and all those things take time. And doing it once per test context is a lot cheaper than doing it twice per test.

AlonaKaplan replied:
I guess I miss something. But here in this specific context I don't see you are creating share storage for the context. The BeforeAll is only calling the cleanup. Isn't it?

maiqueb replied:
I mistakenly assumed you meant this approach on the entire file, not limited to the VM tests. 

You are right; I will take the `BeforeTestsCleanup` route.

EdDev replied:
I have to agree here with Alona, we cannot use the teardown cleanup approach in general per the info collection on failure.
I do not know which is worse really... not having logs or leaking resources.

maiqueb replied:
So let's agree, because the current implementation leaks resources and logs more than it should (for instance, the 3 VMI tests are executed sequentially without tearing down any resources, since the `BeforeTestsCleanup` is executed **only** on the `tests.BeforeAll` clauses ). 

Any alternative is better than what we have right now (e.g. if it currently logs too much and leaks resources, just leaking resources is better).

EdDev commented on tests/expose_test.go:
```diff
@@ -663,10 +656,8 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
-				svc, err := virtClient.CoreV1().Services(vm.Namespace).Get(serviceName, k8smetav1.GetOptions{})
-				Expect(err).ToNot(HaveOccurred())
-				serviceIP := svc.Spec.ClusterIP
+				serviceIP, err := getServiceIP(virtClient, vm.GetNamespace(), serviceName)
```
Can we pass the service object instead? Then there will be no need to fetch the namespace from another object, it is confusing.

maiqueb replied:
Removed these functions.

maiqueb reviewed (commented):
Addressed (most of) @AlonaKaplan's comments.

@EdDev / @AlonaKaplan I just want to point out that the current status quo on the expose tests is that the services are not being cleaned up per context; the jobs are deleted when the TTL expireres; the VMI/VMIRS/VM are deleted as the services are: on the next test `BeforeAll`.

I can preserve that behavior, by following @AlonaKaplan's suggestion of adding a random string to the service name.

This also means that when we get logs, we get 'too many' logs (from all tests in the context).

maiqueb commented on tests/expose_test.go:
```diff
@@ -531,3 +655,10 @@ func resolveNodeIPAddrByFamily(virtClient kubecli.KubevirtClient, sourcePod *k8s
 	return resolveNodeIp(virtClient, sourcePod, *hostname, ipFamily)
 }
+
+func getVirtctlExposeCommand(currentArgs []string, ipFamily k8sv1.IPFamily) []string {
```
My intention is to have a function that gets me the correct command to expose a service over the desired IP family. Would  `generateVirtctlExposeServiceCommand` be OK ? If not, I'll just remove the function, so we can move forward.

maiqueb replied:
Done (removed the function).

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -330,38 +330,79 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+			})
+
+			AfterEach(func() {
+				Expect(job).NotTo(BeNil(), "a successful test must have stored a way to delete the batchv1.Job entity")
```
yes, that's what I meant:)

maiqueb commented on tests/expose_test.go:
```diff
@@ -184,13 +293,28 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 					Expect(node.Status.Addresses).ToNot(BeEmpty())
 					nodeIP := node.Status.Addresses[0].Address
```
Right. Hope it is better now.

maiqueb commented on tests/expose_test.go:
```diff
@@ -184,13 +293,28 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 					Expect(node.Status.Addresses).ToNot(BeEmpty())
 					nodeIP := node.Status.Addresses[0].Address
 
+					if ipFamily == k8sv1.IPv6Protocol {
```
This follow the current status quo of the test and the scope of this PR is adding ipv6 family support to the expose_tests, not fix whatever readability issues the test might have.

While you might be right, I think those fixes should be addressed in a separate PR.

Never-the-less, added a commit which attempts to make the code easier to follow.

maiqueb replied:
@EdDev / @AlonaKaplan come to terms over the direction where this will go https://github.com/kubevirt/kubevirt/pull/4272#discussion_r512165608

maiqueb replied:
I chose to appease the maintainer.

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -179,15 +311,30 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
-					nodeIP := node.Status.Addresses[0].Address
+
+					nodeIP, err := getNodeInternalIPByFamily(node, ipFamily)
+					if err != nil {
```
Please don't assume the err is not find the ipv6 address. In case of ipv4 add regular error handling (assertion). In case of ipv6 do the workaround.

maiqueb replied:
Removed the commit that introduced this.

maiqueb commented:
/retest

maiqueb commented on tests/expose_test.go:
```diff
@@ -536,47 +536,99 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+				}
+			})
+
+			table.DescribeTable("[test_id:1538][label:masquerade_binding_connectivity]Connect to ClusterIP service that was set when VM was offline.", func(ipFamily k8sv1.IPFamily) {
```
Done.

maiqueb commented on tests/expose_test.go:
```diff
@@ -84,34 +116,72 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+			})
+
+			AfterEach(func() {
+				Expect(job).NotTo(BeNil(), "a successful test must have stored a way to delete the batchv1.Job entity")
```
I've read this multiple times and didn't understand what you want . 

If I've reached this point and don't have a job, something went wrong. Period. Hence the expectation, which is the correct way of describing such a relationship between an event and a failed test.

oshoval replied:
isnt it weird that we Expect stuff that the test should verify in its AfterEach ?

maiqueb replied:
I have no idea what you mean.

Never the less, this function is gone, so I guess there's nothing to disagree about.

maiqueb commented on tests/expose_test.go:
```diff
@@ -73,6 +76,35 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+		return err
+	}
+
+	cleanupJob := func(jobName string, namespace string) error {
```
> * Consider adding to `tests/job.go`

I'd rather not derail this PR further by attempting to define whatever API for job cleanups. I have a simple need in this module, which is addressed by the following code. From [this discussion](https://github.com/kubevirt/kubevirt/pull/4272#discussion_r511586890), it is not even clear we want to cleanup jobs, since, they kind of recycle themselves.

> * I would prefer a function signature that gets a pointer to a job and retrieves from it the name and namespace.

I can compromise, since I see those 2 things as being fundamentally the same. End of the day, I don't need the object to delete, just the name and the namespace. Which I retrieve from the object. Whatever.

> * It may be even nicer to have only one function that accepts a variable number of jobs for deletion.

It is implemented below.

> * Do you really want to use the teardown pattern given what we know?

Given that the test is currently worse than that (resources leak throughout the tests, which mean that when anything gets logged, it logs the resources of more than one test), yes. I think right now what we have is worse.

AlonaKaplan replied:
If I need to debug a failing test. I would prefer extra (even non relevant logs) over no logs at all.

maiqueb replied:
Stopped removing the resources. We're now back to using the previous approach, where the entities we create get deleted once per context, in the `tests.BeforeAll` .

maiqueb commented on pkg/virtctl/expose/expose_test.go:
```diff
@@ -238,6 +238,27 @@ var _ = Describe("Expose", func() {
 				Expect(err()).To(BeNil())
 			})
 		})
+		Context("With IPv4 IPFamily", func() {
```
> Can you also check the default case?

Yes, I can.

> 
> BTW, I do not understand why two levels have been used here, having 1 context on 1 It makes the context redundant IMO.

Take a look at the rest of the test; it is the pattern currently set in place. And I don't volunteer to change that.

> Sounds more reasonable to "group" the "With IP family" and have multiple It/s under.

Logically, it is what makes sense, I agree. Unfortunately, none of the tests on that entire file follow those rules.

maiqueb replied:
A successful test w/ the default family is presented in https://github.com/kubevirt/kubevirt/blob/839e1784a370bd515462d98494ea52f8d4f9d496/pkg/virtctl/expose/expose_test.go#L127

maiqueb commented on tests/expose_test.go:
```diff
@@ -73,6 +76,35 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+	}
+
+	cleanupJobs := func(jobs []batchv1.Job) []error {
+		var errorBucket []error
```
Removed this method.

maiqueb commented on tests/expose_test.go:
```diff
@@ -73,6 +76,35 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 		return job
 	}
 
+	cleanupService := func(serviceName string, namespace string) error {
```
Function is gone.

maiqueb commented on pkg/virtctl/expose/expose.go:
```diff
@@ -125,6 +128,17 @@ func (o *Command) RunE(cmd *cobra.Command, args []string) error {
 		return fmt.Errorf("unknown service type: %s", strServiceType)
 	}
 
+	// convert from string to IPFamily type enum
```
Done.

maiqueb commented:
/retest

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -491,3 +493,39 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+	if err != nil {
+		return "", err
+	}
+	for _, ipStr := range strings.Split(output, "\n") {
```
Why do you have a for loop?

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -298,40 +298,75 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
-			const serviceName = "cluster-ip-udp-vmi"
+			const serviceNamePrefix = "cluster-ip-udp-vmi"
+
+			var job *batchv1.Job
```
Now you don't handle the job removal, I think you can remove the declaration from here.

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -298,40 +298,75 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
-			const serviceName = "node-port-udp-vmi"
+			const serviceNamePrefix = "node-port-udp-vmi"
+
+			var jobs []batchv1.Job
```
same

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -720,8 +745,7 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 				By("Deleting the VM.")
-				Expect(virtClient.VirtualMachine(vm.Namespace).Delete(vm.Name, &k8smetav1.DeleteOptions{})).To(Succeed())
-				tests.WaitForVirtualMachineToDisappearWithTimeout(vmi, 120)
+				Expect(cleanupVM(virtClient, vm.GetNamespace(), vm.GetName())).To(Succeed(), "the VM object should be deleted")
```
Why is this special cleanup needed? The test is flaky if we don't wait for the vm object removal?

maiqueb reviewed (commented):
Replied to @AlonaKaplan / @EdDev 's comments.

maiqueb commented:
/retest

maiqueb commented:
/retest

EdDev commented on tests/expose_test.go:
```diff
@@ -72,6 +75,10 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 		return job
 	}
 
+	generateUniqueServiceName := func(serviceName string) string {
```
I did not know we have such a practice to define things in the smallest scope (at least not in our own functions).
Such functions may mistakenly use an outer variable and therefore be dependent unexpectedly on other side effects.
(I think I saw it happening in this PR already)

I would expect the name to be `ramdomName()`.

EdDev commented on tests/expose_test.go:
```diff
@@ -507,50 +508,86 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
+			return vmi
+		}
+
+		createStoppedVM := func(virtClient kubecli.KubevirtClient, namespace string) (*v1.VirtualMachine, error) {
```
Some of the functions use the outer `virtClient` (startVMIFromVMTemplate), some do not.
In the same context, why you pass in the namespace? I do not see now anything else using it with a different input.

I guess this is a repeated pattern here. If the lambda is not either inlined or using outer values to construct it, I find it very confusing (and requires from my part to lookup for which vars you do use).

maiqueb replied:
I understand that lambdas that use stuff defined out of their body is extremely confusing, and reading that code without an IDE a nightmare - that's why I make a serious effort of passing the attributes to the functions. Please assume all variables must be passed as arguments, and keep letting me know when they don't.

EdDev replied:
Well, that rule is just contradicting the existence of lambdas in Go IMO.
I do not think they exist to limit scope, they exist to either do something inlinned or use an outer variable in order to carry a dynamic state (i.e. it is a generated function and not a static one).

EdDev commented on tests/network/services.go:
```diff
@@ -44,8 +44,13 @@ import (
 
-	runTCPClientExpectingHelloWorldFromServer := func(host, port, namespace string) *batchv1.Job {
-		job := tests.NewHelloWorldJob(host, port)
+	runTCPClientExpectingHelloWorldFromServer := func(host, port, namespace string, isIPv6 bool) *batchv1.Job {
```
I do not like the usage of booleans, especially when they go manly levels in the stack.
I have hard time to follow and suggest a better approach, so I guess I give up.

maiqueb replied:
In this case, I do not have a much better alternative, since `host` features an FQDN, and not an ip address.

I can duplicate the function having `runTCPClientExpectingHelloWorldFromServer` and `runTCPClientExpectingHelloWorldFromServerIPv6` if you prefer though.

EdDev replied:
We will still be left with at least one func that has a boolean as argument, so I do not think it is worth it.

EdDev reviewed (approved):
Thank you!
My remaining comments are not worth blocking.

maiqueb commented:
Added the `kubecli.KubevirtClient` to the VM lambdas where it was missing and updated the name of the function that randomizes the string, as per @EdDev 's request.

Thanks for the thorough review.

AlonaKaplan commented on tests/job.go:
```diff
@@ -129,19 +139,39 @@ func NewHelloWorldJobUDP(host string, port string) *batchv1.Job {
 
-	return job
+// NewHelloWorldJobHTTPv6 is a workaround to generate `batch.V1` jobs that will access services exposed via IPv6 addresses.
+// It does exactly what NewHelloWorldJobHTTPv6 does, but for IPv6.
```
s/NewHelloWorldJobHTTPv6/NewHelloWorldJobHTTPv

AlonaKaplan commented on tests/job.go:
```diff
@@ -111,9 +111,19 @@ func NewJob(name string, cmd, args []string, retry, ttlAfterFinished int32, time
+
+// NewHelloWorldJobv6 is a workaround to generate `batch.V1` jobs that will access services exposed via IPv6 addresses.
+// It does exactly what NewHelloWorldJob does, but for IPv6.
+func NewHelloWorldJobv6(host string, port string) *batchv1.Job {
```
I would also avoid the duplication between the ipv4 and ipv6 versions.

For example -
```
func newHelloWorldJob(host string, port string, ping boolean) *batchv1.Job {
        var pingCmd string
        if ping {
            pingCmd = fmt.Sprintf("ping -c 1 %s;", host)
        }
}
	check := []string{fmt.Sprintf(`set -x; %s x="$(head -n 1 < <(nc %s %s -i 3 -w 3))"; echo "$x" ; if [ "$x" = "Hello World!" ]; then echo "succeeded"; exit 0; else echo "failed"; exit 1; fi`, pingCmd, host, port)}
	job := NewJob("netcat", []string{"/bin/bash", "-c"}, check, JobRetry, JobTTL, JobTimeout)
	return job
}

func NewHelloWorldJob(host string, port string) *batchv1.Job {
	newHelloWorldJob(host, port, false)
}

func NewHelloWorldJob(host string, port string) *batchv1.Job {
	newHelloWorldJob(host, port, true)
}

```

maiqueb replied:
I prefer to avoid using the boolean parameter, and duplicate the functions instead - since a boolean argument implicitly makes a function do one of 2 things, when it is supposed to do a single one.

Hope this is not a deal breaker .

AlonaKaplan replied:
You can pass a string with a `preCmd` instead. I really don't like the duplication.
But I won't insist.

AlonaKaplan commented on tests/expose_test.go:
```diff
@@ -322,56 +507,96 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 			By("Getting the status of the VMI")
 			Eventually(func() bool {
-				vm, err = virtClient.VirtualMachine(vm.Namespace).Get(vm.Name, &k8smetav1.GetOptions{})
+				vm, err = virtClient.VirtualMachine(namespace).Get(name, &k8smetav1.GetOptions{})
```
You asked for a notice when you use general variable from a method:)

maiqueb replied:
Good catch! Sneaky side effect ...

AlonaKaplan commented:
Thanks for the fixes, I know the review process wasn't easy:)
/approve
/hold - there are some small comments, feel free to remove the hold once they are fixed.

maiqueb commented on tests/expose_test.go:
```diff
@@ -82,13 +89,30 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 			const servicePort = "27017"
-			const serviceName = "cluster-ip-vmi"
-			It("[test_id:1531][label:masquerade_binding_connectivity]Should expose a Cluster IP service on a VMI and connect to it", func() {
+			const serviceNamePrefix = "cluster-ip-vmi"
```
I think it improves readability (for instance, I have the understanding that the prefix is preserved without reading the method):

```golang
# I think this reads better
serviceName = randomizeName(serviceNamePrefix)

# than this
serviceName = randomizeName("blablalbla")
```

I can make it a var if you insist, but I see no reason to allow mutation on it.

maiqueb reviewed (commented):
Addressed the final comments. 

Removing the hold - this still needs to fight CI and get a new lgtm tag.

/hold cancel

EDIT: I do realize how futile it is to write 'addressed the **final** comments' :P

EdDev reviewed (approved):
Thank you!

maiqueb commented:
Addressed the last of @AlonaKaplan 's comments; hope all is well now.

PS: I think it does look better without the duplicated code.

maiqueb commented:
/retest
